### PR TITLE
Adding HMAC and ECC peripheral support

### PR DESCRIPTION
**Pull request by nabilmerk**
This is the project version for the HMAC driver. It's been tested only on ESP32C3 using a zero key on efuse.
I'd like to hear your feedback on this PR.

### CODE REVIEW

File: esp-hal-common/src/utils.rs
Comment (MabezDev) [inline esp-hal-common/src/utils.rs:1]: Not a huge fan of the "utils" name, could we rename this `reg` or `reg_access`, or something like that? @esp-rs/espressif feel free to suggest names too.

File: esp-hal-common/src/utils.rs
Comment (MabezDev) [inline esp-hal-common/src/utils.rs:24]: I believe this comes from `sha.rs`, can we remove it from there if it exists here?

File: esp32c3-hal/examples/hmac.rs
Comment (MabezDev) [inline esp32c3-hal/examples/hmac.rs:57]: This should be done in the driver constructor, I know your PR is for HMAC but if you could fix it for SHA at the same time that would be appreciated!

**Review by MabezDev — CHANGES_REQUESTED** [review]
Thanks for the PR, this is looking good! Just a few comments and the CI to fix :)

File: esp-hal-common/src/utils.rs
Comment (nabilmerk) [reply esp-hal-common/src/utils.rs:1]: I wasn't a big fan too of the name so I'll go with your suggestions.

File: esp32c3-hal/examples/hmac.rs
Comment (nabilmerk) [reply esp32c3-hal/examples/hmac.rs:57]: Just for my knowledge, is there any specific reason that enabling should be done in the driver constructor ?
It make sense that the clock gating and reset logic should be handled "elsewhere" because it can lead to over consuming in embedded software or am I wrong ?

File: esp32c3-hal/.cargo/config.toml
Comment (bjoernQ) [inline esp32c3-hal/.cargo/config.toml:2]: In the whole repo we use the `espflash 2.x` syntax already - so would you mind chaning this back to `espflash flash --monitor`

I know the situation regarding `espflash 1.x` vs `espflash 2.x` is quite annoying and confusing but at least in one repo we should be consistent

File: esp32c3-hal/examples/hmac.rs
Comment (bjoernQ) [reply esp32c3-hal/examples/hmac.rs:57]: I think it's a bit of a convention in this HAL and we weren't very consistent about this in the past.
For me the idea is to have the driver in a working state after constructing it. There would be other ways to ensure the user enabled the peripheral clock but those would be more complex.
Also, probably in future we should make sure to disable the peripheral clock when the driver gets dropped 🤔 

btw I really like the detailed documentation in this PR 👍

File: esp-hal-common/src/utils.rs
Comment (nabilmerk) [reply esp-hal-common/src/utils.rs:24]: OK and I will adapt `sha.rs` for calling it.

File: esp32c3-hal/.cargo/config.toml
Comment (nabilmerk) [reply esp32c3-hal/.cargo/config.toml:2]: I was using `espflash 1.x` so it wasn't needed and I push it in the commit unintentionaly. I'll remove it.

File: esp-hal-common/src/utils.rs
Comment (nabilmerk) [reply esp-hal-common/src/utils.rs]: Done in the commit.

File: esp32c3-hal/examples/hmac.rs
Comment (nabilmerk) [reply esp32c3-hal/examples/hmac.rs]: Updated for HMAC and SHA. I also updated examples.

**Comment by MabezDev** [comment]
Thanks for making those changes, looking better already :). I tried running some of the examples, but sadly they seem to panic with "KeyPurposeMismatch", could you look into that?

**Comment by nabilmerk** [comment]
Have you set the key in the efuse block ? You need to set the key purpose as well in the efuse block and modify the example with the corresponding key for the SW.

**Comment by nabilmerk** [comment]
I've tested on my local repo on github to make sure that all workflows works with the last commit : all passed.

**Comment by MabezDev** [comment]
Thanks for getting the CI green!

> Have you set the key in the efuse block ? You need to set the key purpose as well in the efuse block and modify the example with the corresponding key for the SW.

Oops, working now! Sorry for the noise :D.

This is looking good! Could you rebase the changes to resolve the conflicts?

**Comment by bjoernQ** [comment]
I can reproduce this with `rustc 1.70.0-nightly (4087deacc 2023-04-12)` while that didn't happen with e.g. `rustc 1.70.0-nightly (39f2657d1 2023-03-09)`

**Comment by nabilmerk** [comment]
I believe the issue is coming from `rustc` and the `embedded-hal-async` library. What are the next steps to move forward?

**Comment by bjoernQ** [comment]
Problem with CI is that the latest nightlies refuse to compile embedded-hal-async - I will create a PR to pin the nightly compiler version in CI for now (until embedded-hal-async got fixed) which should make CI happy again

**Comment by bjoernQ** [comment]
Rebasing should fix the CI problem (hopefully)

**Comment by nabilmerk** [comment]
The rebase has fixed the CI in my repo.

File: esp32c2-hal/Cargo.toml
Comment (jessebraham) [inline esp32c2-hal/Cargo.toml:46]: I don't think these are being used.

File: esp32c3-hal/Cargo.toml
Comment (jessebraham) [inline esp32c3-hal/Cargo.toml:52]: Don't think this is needed, either.

**Review by jessebraham — CHANGES_REQUESTED** [review]
Looks pretty good to me, thanks for the PR. I left a couple notes regarding dependencies, and there are also a number of warnings depending on which package is being built; if you could resolve these please that would be great!

File: esp32c2-hal/Cargo.toml
Comment (nabilmerk) [reply esp32c2-hal/Cargo.toml:46]: You're right! I've started coding for ECC and these lines are not necessary for HMAC :).
I'll update the file to remove them.

File: esp32c3-hal/Cargo.toml
Comment (nabilmerk) [reply esp32c3-hal/Cargo.toml:52]: I was using them before settling stress test so I will remove it.

**Comment by nabilmerk** [comment]
All the warnings regarding HMAC and SHA have been cleaned.

**Comment by nabilmerk** [comment]
When this PR will be merged, I will be adding another PR regarding ECC accelerator.

**Comment by bjoernQ** [comment]
Looks pretty good to me but since there were also some comments by others let's wait for their review

**Review by bjoernQ — APPROVED** [review]

**Comment by nabilmerk** [comment]
I've added ECC peripheral support as well and change the name of the PR.
@jessebraham BTW, when I was testing with the esp32c2 board, I had to force the baudrate for monitoring to 74880 because my board has a crystal of 26 MHz. I'm using the version 2.0.0-rc.3 of espflash which has the improvements relating to crystal frequency and serial monitor for the ESP32-C2. Is it a normal behavior to force the baudrate or should I open an issue in espflash ?

**Comment by bjoernQ** [comment]
I think to get things merged it would be better to split the PR by peripherals. That would make reviewing much easier and it's better to have one merge per peripheral.

**Comment by jessebraham** [comment]
Thanks @bjoernQ for bumping this, apologies to @nabilmerk that we've left this to languish for so long.

As already stated, this PR got quite large and deals with multiple peripherals, making it rather difficult to review and test. Breaking this up would make things much easier on our end.

The way I see it this could probably be 3 separate PRs:

- Refactoring the `SHA` driver and alignment helper
- Adding the `HMAC` driver
- Adding the `ECC` driver

**Comment by AnthonyGrondin** [comment]
Hi. What is the state of this PR? I started working on adding the HMAC driver, before seeing that this PR was opened. Are there any changes needed before merging?

I'm looking forward to HMAC hardware acceleration to be implemented, to implement the digital signature (DS) driver.

**Comment by jessebraham** [comment]
@AnthonyGrondin If you want to do a full review of this PR and test each example for each chip you're welcome to, then we could possibly move forward. I don't have the time or energy to do that, because it will take hours probably, which is why we asked for this to be split up.

**Comment by jessebraham** [comment]
@AnthonyGrondin did you have any intention of tackling the `ECC` peripheral as well? No worries if not, just curious. Somebody else (or me) can pick it up if you do not plan on doing so. Just would like to get this closed 😁

**Comment by AnthonyGrondin** [comment]
Hi,

No, I didn't have any intentions of implementing the `ECC` peripheral.

**Comment by AnthonyGrondin** [comment]
I think:
- ECC
- Digital Signature (DS)
- XTS_AES

are the only cryptographic accelerators left that haven't been implemented yet. I mostly did HMAC so that I could implement DS but I found that I might not need it for what I intended to do, so I haven't worked on DS much, except from a mere WIP... If someone wants to do it :grin:

**Pull request closed without merging**
